### PR TITLE
Add the system verification test to the kubeadm preflight checks

### PR DESCRIPTION
**Pull request by luxas**
And refactor the system verification test to accept to write to a specific writer in order to customize the output

This PR is targeting v1.5, PTAL
cc @Random-Liu @dchen1107 @kubernetes/sig-cluster-lifecycle

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36334)

### CODE REVIEW

**Comment by Random-Liu** [comment]
@luxas I didn't consider that we have the requirement to print result to specific stream.

If so, I'd prefer add a reporter, define the output stream in it, and pass it when creating the validators.

However, this may need a bit more code refactoring. For now, if you want to make the change minimal, we can have a global reporter, and let all validators use it for now.

**Comment by Random-Liu** [comment]
```
I1106 19:13:43.701] FAILED   hack/make-rules/../../hack/verify-gofmt.sh	9s
```

And do you have time to address the comment today?
If not, I can also send a PR to refactor the system verification a bit to make it easier to you to use. Is that OK for you?

**Comment by luxas** [comment]
I think I have something that's really close now.
I can upload it now, although I don't think it fully works 

I'm on the sig-cluster-lifecycle meeting now at Google Seattle...

**Comment by Random-Liu** [comment]
@luxas ACK. Sure, then will wait for your PR. :)

**Comment by luxas** [comment]
@Random-Liu I'm pretty sure this doesn't work :). PTAL
It seems like the output isn't written to the writer... I guess that's because the writer is copied by value instead of reference somehow.

You can test it like this:
```
package main

import (
	"fmt"
	"os"
	"bufio"
	"k8s.io/kubernetes/test/e2e_node/system"
)

func main() {
	bufw := bufio.NewWriterSize(os.Stdout, 1 * 1024 * 1024)
	err := system.Validate(system.DefaultSysSpec, &system.StreamReporter{WriteStream: bufw})
	fmt.Printf("err: %v\n", err)
	bufw.Flush()
}
```

And it should print out the output of the system verification test, but it doesn't.
Please review and say what I'm missing

**Comment by Random-Liu** [comment]
```
system.StreamReporter does not implement system.Reporter (Report method has pointer receiver)
```

**Comment by Random-Liu** [comment]
@luxas 
This works, the problem is in the code:
https://github.com/kubernetes/kubernetes/pull/36334/files#diff-e5e568565e3ad86dadb2141ff07b0ffbR66
`WriteStream != nil` => `WriteStream == nil`.

File: test/e2e_node/system/validators.go [hunk 1]
```diff
@@ -29,21 +28,31 @@ type Validator interface {
...
+	// validators are all the validators.
+	var validators = []Validator{
+		&OSValidator{Reporter: report},
+		&KernelValidator{Reporter: report},
+		&CgroupsValidator{Reporter: report},
+		&DockerValidator{Reporter: report},
+	}
+
 	for _, v := range validators {
-		glog.Infof("Validating %s...", v.Name())
```
Comment (Random-Liu) [inline test/e2e_node/system/validators.go]: Why remove this? And same for above.

File: test/e2e_node/system/types.go [hunk 2]
```diff
@@ -120,3 +124,8 @@ var DefaultSysSpec = SysSpec{
 		},
 	},
 }
+
+// DefaultReporter is the default Reporter
+var DefaultReporter = StreamReporter{
```
Comment (Random-Liu) [inline test/e2e_node/system/types.go]: Move this to `report.go`.

File: test/e2e_node/system/report.go [hunk 3]
```diff
@@ -58,5 +63,10 @@ func report(item, r string, t resultType) {
 	default:
 		c = white
 	}
-	fmt.Printf("%s: %s\n", wrap(item, white), wrap(r, c))
+	if dr.WriteStream != nil {
```
Comment (Random-Liu) [inline test/e2e_node/system/report.go]: `dr.WriteStream == nil`.

File: test/e2e_node/system/os_validator_test.go [hunk 4]
```diff
@@ -17,13 +17,16 @@ limitations under the License.
 package system
 
 import (
+	"os"
```
Comment (Random-Liu) [inline test/e2e_node/system/os_validator_test.go]: Why `os`?

File: test/e2e_node/system/kernel_validator_test.go [hunk 5]
```diff
@@ -18,13 +18,16 @@ package system
 
 import (
 	"bytes"
+	"os"
```
Comment (Random-Liu) [inline test/e2e_node/system/kernel_validator_test.go]: Why `os`?

**Comment by luxas** [comment]
Thanks @Random-Liu!
Oh how silly doesn't that feel? :smile:
First had the `dr.WriteStream != nil` check to actually do the right thing, but then changed the logic without changing the if condition. Feels quite stupid now :), however I was doing a lot if things simultaneously while uploading this during the sig meeting.

Will fix and get this hopefully lgtm'd today

**Comment by Random-Liu** [comment]
@luxas Sure. Will review again as soon as you update the code. :)

**Comment by luxas** [comment]
@dchen1107 @Random-Liu PTAL, updated
I guess we want it somehow lgtm'd before midnight.

**Comment by luxas** [comment]
Re the glog stuff: it's just that I don't see any use for glog here...
I mean, it should be used when you have a component that's logging different kinds of stuff all the time, but not here when we specify a reporter. Now all output clearly should be given to the reporter that then decides what to do with that.

IMO it looks cleaner to not mix fmt and glog output and also kubeadm isn't using glog :)
OK as-is?

**Comment by Random-Liu** [comment]
> Re the glog stuff: it's just that I don't see any use for glog here...

The things printed by glog are not validation result, they are for debugging and testing, and we may need to add more glog to report temporary error or something in the future.
When user really uses the application, we won't set `--alsologtostderr` flag of glog, so that user will not see anything from glog.

> kubeadm isn't using glog

Many packages imported by kubeadm is using glog, such as `pkg/cloudprovider/providers`. The system verification is just one of those packages, I don't see any reason that we can't use glog for debugging purpose in system verification package.

**Comment by luxas** [comment]
But are you okay with not using glog (or are you blocking on it)?
I'm not sure I know how to update bazel, it failed locally for me (cc @mikedanese)
Can I do it manually?

Anyway, if the design is lgtm to you, I guess it's considered valid for v1.5 (cc @saad-ali), because the only thing that's blocking the lgtm is the bazel stuff

**Comment by Random-Liu** [comment]
Just run `hack/update-bazel.sh` :)

**Comment by Random-Liu** [comment]
@luxas I prefer to add the glog back, because it's useful for future debugging. :)

**Comment by saad-ali** [comment]
> And I also hope we can get this in for v1.5.

Is this a bug fix? Removing 1.5 milestone until verification.

**Comment by Random-Liu** [comment]
```
hack/make-rules/../../hack/verify-bazel.sh	25s
```
Could you rebase and run `hack/update-bazel.sh` again?

**Comment by luxas** [comment]
@Random-Liu Everything passed, PTAL

**Comment by Random-Liu** [comment]
LGTM, thanks @luxas.

This is a kubeadm v1.5 feature. The PR is almost ready on release day. (See https://github.com/kubernetes/kubernetes/pull/36334#issuecomment-259065990)

/cc @saad-ali How should we deal with this?

**Comment by luxas** [comment]
@saad-ali I'd love to get it into v1.5 if possible, but I do not strictly block on it.

Regarding kubeadm, we've already "released" our stable v1.5 version from master, so it do not explicitely have to merge into v1.5 for kubeadm to be stable, but it would definitely help the further development since we will split kubeadm out of the main repo and then this refactoring in the main k8s repo is essential.

TL;DR; This is more a "nice-to-have", but since it was basically lgtm'd on the code freeze day and will help the further kubeadm work, it would be great to have merged if possible.

**Comment by saad-ali** [comment]
> TL;DR; This is more a "nice-to-have", but since it was basically lgtm'd on the code freeze day and will help the further kubeadm work, it would be great to have merged if possible.

Since it is non-critical/non-bug fix. Let's leave it out of 1.5. We don't want to risk destabilizing the release with unnecessary new functionality post-code freeze.

**Comment by Random-Liu** [comment]
> Since it is non-critical/non-bug fix. Let's leave it out of 1.5. We don't want to risk destabilizing the release with unnecessary new functionality post-code freeze.

@saad-ali SGTM.

@luxas Let's get this merged as soon as the code freeze is over. :)

**Comment by rmmh** [comment]
@k8s-bot test this

**Comment by dims** [comment]
@k8s-bot cvm gke e2e test this

**Comment by dims** [comment]
@k8s-bot gci gke e2e test this

**Comment by saad-ali** [comment]
Removing 1.5 milestone (it was readded by mistake)

File: cmd/kubeadm/app/preflight/checks.go [hunk 6]
```diff
@@ -213,9 +215,26 @@ func (hst HttpProxyCheck) Check() (warnings, errors []error) {
 	return nil, nil
 }
 
+type SystemVerificationCheck struct{}
+
+func (sysver SystemVerificationCheck) Check() (warnings, errors []error) {
+	// Create a buffered writer and choose a quite large value (1M) and suppose the output from the system verification test won't exceed the limit
+	bufw := bufio.NewWriterSize(os.Stdout, 1*1024*1024)
```
Comment (deads2k) [inline cmd/kubeadm/app/preflight/checks.go]: Are you ever actually using the data you're collecting?  If not, it seems like `io.Discard` is what you're looking for.  If you are, why not create a writer which keeps the first and the last and elides the middle if too much is written instead of hoping you have enough space?
